pkg/controllers/gateway/gateway_controller.go

```diff
@@ -180,9 +185,31 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	return ctrl.Result{}, reconcileErr
 }
 
+func (r *GatewayReconciler) reconcileDefaultDNSPolicy(ctx context.Context, upstreamGateway *gatewayv1beta1.Gateway) error {
+	log := crlog.FromContext(ctx)
+
+	defaultDNSPolicy := policy.NewDefaultDNSPolicy(upstreamGateway)
+
+	var dnsPolicies v1alpha1.DNSPolicyList
+	err := r.List(ctx, &dnsPolicies, client.InNamespace(upstreamGateway.GetNamespace()), client.MatchingFields{policy.POLICY_TARGET_REF_KEY: policy.GetTargetRefValueFromPolicy(&defaultDNSPolicy)})
```

Comment:
If I am not wrong, this operation is doing: "read all the dns policies that are targeting this gateway". 
I think that you do not need to do this if you rely on the annotations in the gateway which are back references to the DNS policies. In other words, Just check gateway annotation for a back references.

Reply:
Yes, this is listing the polices in the current namespace that are targeting this gateway. Whether or not we can use the annotations instead relates to this question https://github.com/Kuadrant/multicluster-gateway-controller/pull/237#discussion_r1226412200 I'm not really seeing how we can reliably, but maybe I'm wrong.

Reply:
I think that this operation is only needed for an edge case: First you create some DNS policy targeting a gateway that does not exist. Then you create the gateway. In that scenario, the gateway would not have any finalizer annotation and `ensureDefaultDNSPolicy` method would be called. Then, this operation would detect that policy and prevent the creation of the default dns policy. 

For the normal use case, when the gateway has been created and it does not have the kuadrant's finalizer, the default DNS policy needs to be created. Then the user needs to delete it and create his own one. 

I do not expect this edge situation to happen often, but yeah, who knows. LGTM